[WIP] CM-281,280,282,287,286,283,285

e-molochko opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ozalexo/chronomint-rn/36)
<!-- Reviewable:end -->

e-molochko committed:
Changed login algorithm + dynamic navigation after eth wallet create

e-molochko committed:
temp request commit

ozalexo commented on packages/session/redux/thunks.js:
```diff
@@ -4,10 +4,13 @@
   try {
-    dispatch(login(ethAddress))
+    dispatch(createBitcoinWallet(privateKey, ethAddress))
+      .then(dispatch(login(ethAddress)))
```
Does it work?

I expect something like this here:
```
.then(() => dispatch(login(ethAddress)))
```

ozalexo commented on src/screens/Login/ConfirmMnemonic/ConfirmMnemonicContainer.js:
```diff
@@ -43,7 +47,10 @@ class ConfirmMnemonicContainer extends PureComponent {
 
     createAccount(mnemonic, password)
-    navigation.navigate('WalletList')
+      .then(() => {
```
Suggestion:
```
      .then(() => {
        const [account] = accounts.slice(-1) // get last created account
        const page = account ? 'Login' : 'Start' // need to verify that account is falsy here
        navigation.navigate(page, { account })
      })
```

ozalexo commented on src/screens/Login/Login/LoginContainer.js:
```diff
@@ -73,38 +74,43 @@ class LoginContainer extends PureComponent {
-      this.handleLogin(address)
+  handleLoginClick = async ({password}) => {
+    const pass = password ? password : this.state.password
+    const results = await this.checkPassword(pass)
```
Please rename 'results' into something like 'isPasswordValid'

ozalexo commented on src/screens/Login/Login/LoginContainer.js:
```diff
@@ -73,38 +74,43 @@ class LoginContainer extends PureComponent {
+        navigate('WalletList')
+      })
+      .catch((error) => {
+        console.log('Login failure:', error)
```
We should notify user about this. Alert is OK for now.

ozalexo commented on packages/ethereum/redux/thunks.js:
```diff
@@ -4,27 +4,23 @@
 
-    dispatch(bitcoinCreateWallet(ethAddress, bitcoinAddress))
-    dispatch(login(ethAddress))
+    dispatch(savePrivateKey(decryptedWallet.privateKey))
```
Thunk createAccount must create account. We should not saveProvateKey inside this thunk.
We need to call `dispatch(savePrivateKey(decryptedWallet.privateKey))` right after call createAccount thunk in appropriate place (we will discuss it)

e-molochko committed:
Added requests
- Add request for BTC wallet ballance
- Add request for subscription on BTC ballance updates
- Little fixes in login thunks

e-molochko committed:
temp requests commit

e-molochko committed:
network fixes + bignumber.js for crypto ballance

e-molochko committed:
Added update of btc wallet ballance(need change some ballance handling)

ozalexo commented on packages/network/middlewares/rabbitmq/RmqManager.js:
```diff
@@ -81,9 +81,9 @@ class RmqManager {
-      if (this.client && !this.client.connected) {
-        return reject('RmqManager subscribe error: no connection to RabbitMQ host')
-      }
+      // if (this.client && !this.client.connected) {
```
Please delete these lines. We will return to it later

ozalexo commented on packages/bitcoin/redux/constants.js:
```diff
@@ -7,6 +7,7 @@ export const DUCK_BITCOIN = 'bitcoin'
 
 
 export const BITCOIN_CREATE_WALLET = 'BITCOIN/CREATE_WALLET'
+export const BITCOIN_UPDATE_BALANCE = 'BITCOIN/BALANCE/UPDATE'
```
Naming conventions here a bit different:
DOMAIN_DUCK/OTHER_PART_OF_CONSTANT
So in this case it should be BITCOIN/BALANCE_UPDATE

ozalexo commented on packages/bitcoin/redux/reducers.js:
```diff
@@ -30,6 +54,14 @@ const mutations = {
           address,
+          tokens: {
+            [BTC_PRIMARY_TOKEN]: {
+              balance: null, // это баланс для UI, то есть amount (ниже) пересчитанный с учетом DECIMALS
```
Please remove or replace comments here. There were just internal information comments

ozalexo commented on patches/redux-axios-middleware+4.0.0.patch:
```diff
@@ -8,7 +8,7 @@ patch-package
 +    function getClientNameFromState (blockchain) {
 +      var state = getState()
-+      var chronobankMiddlewares = state.nodes.selected.chronobankMiddlewares
++      var chronobankMiddlewares = state.network.selected.chronobankMiddlewares
```
This change already in remake branch. Please rebase.

ozalexo commented on src/screens/Login/ConfirmMnemonic/ConfirmMnemonicContainer.js:
```diff
@@ -9,12 +9,15 @@ import { connect } from 'react-redux'
-  return {}
+const mapStateToProps = (state) => {
+  return {
+    accounts: getEthAccounts(state),
```
I suggest to rename it to getEthAccountList
Any selector which returns array must have '*List' postfix.

ozalexo commented on src/screens/Login/ConfirmMnemonic/ConfirmMnemonicContainer.js:
```diff
@@ -43,7 +47,12 @@ class ConfirmMnemonicContainer extends PureComponent {
+      .then(() => {
+        const [account] = accounts.slice(-1) // get last created account
+        const params = account ? { account } : null
+        const page = params ? 'Login' : 'Start' // need to verify that account is falsy here
```
Please remove comment in this line.

ozalexo commented on src/screens/Login/Login/LoginContainer.js:
```diff
@@ -73,38 +74,43 @@ class LoginContainer extends PureComponent {
-      this.handleLogin(address)
+  handleLoginClick = async ({password}) => {
+    const pass = password ? password : this.state.password
+    const passwordValidationResults = await this.checkPassword(pass)
```
passwordValidationResults -> isPasswordValid

e-molochko committed:
PR comments fixes

molochkoego committed:
Added subscribe update handler, also added requests for all BTC wallets

molochkoego committed:
Almost added send form for BTC

molochkoego committed:
Fixed send form, added selectwallet and txdraft reducers,actions,thunks

molochkoego committed:
Added NavigationEvents

ozalexo commented on packages/bitcoin/redux/reducers.js:
```diff
@@ -22,6 +23,82 @@ const bitcoinRehydrate = (state, payload) => {
 const mutations = {
 
   [REHYDRATE]: bitcoinRehydrate,
+  [ActionsTypes.BITCOIN_DELETE_TX_DRAFT]: (state, { address, parentAddress }) => {
```
Let's do some reformatting to improve our debug capabilities. Here we need named functions, like it is implemented at packages/ethereum/middleware/reducers.js

ozalexo commented on packages/bitcoin/redux/actions.js:
```diff
@@ -11,6 +11,25 @@ export const bitcoinTxUpdate = (entry) => ({
   address: entry.tx.from,
   entry,
 })
+export const bitcoinCreateTxDraft = ({ address, parentAddress }) => ({
```
Please append new line between functions.

ozalexo commented on packages/bitcoin/redux/thunks.js:
```diff
@@ -2,3 +2,66 @@
+  }
+}
+
+export const updateBitcoinWalletBalance = ({ address, parentAddress, balance, amount }) => (dispatch) => {
```
It would be better if single-action thunk will have the same name as an action. In this case updateBitcoinWalletBalance -> bitcoinUpdateBalance

From my point of view, search through the whole project will be much more comfortable. And there will be no issues with same names, because we are using only thunks in components.

molochkoego committed:
Added crypto images for send form

molochkoego committed:
Removed console.logs

ozalexo commented on src/screens/Wallet/WalletList/WalletListContainer.js:
```diff
@@ -5,43 +5,110 @@
+              rmqSubscribe({
+                channel: `/exchange/events/internal-testnet-bitcoin-middleware-chronobank-io_balance.${address}`,
+                handler: (data) => {
+                  console.log("data: ...............", data)
```
Hm. This is not bad idea, to subscribe after success call to requestBitcoinSubscribeWalletByAddress, but not so good also... 
If REAST API fail, then we will have a chance to obtain required info via WebSocket. So we need to make two these calls independent. We need to subscribe to updates in any case.

ozalexo commented on packages/bitcoin/redux/selectors.js:
```diff
@@ -39,6 +39,18 @@ export const getBitcoinWallets = createSelector(
 )
 
+export const getBitcoinWallets = createSelector(
+  getDuckBitcoin(),
```
We need to make `getDuckBitcoin` like this: 
```
// store.market
export const getDuckBitcoin = (state) =>
  state[DUCK_BITCOIN]
```
In this case we do not need to use () here.


I know, that it is copy-paste from ethereum selector, I'll fix it there also.

ozalexo commented on packages/bitcoin/redux/thunks.js:
```diff
@@ -2,3 +2,66 @@
+export const selectBitcoinWallet = ({ address }) => (dispatch) => {
+  try {
+    dispatch(bitcoinSelectWallet(address))
+    return Promise.resolve()
```
Today I've found that such code may produce errors like `Unhandled promise`. Let's change it to 
```
export const anyThunk = () => (dispatch) => {
  return new Promise((resolve, reject) => {
    try {
      dispatch(anyAction())
      return resolve()
    } catch (error) {
      return reject(error)
    }
  })
}
```

ozalexo commented on packages/bitcoin/utils/amount.js:
```diff
@@ -0,0 +1,33 @@
+ import BigNumber from 'bignumber.js'
+ import web3utils from 'web3/lib/utils/utils'
+
+export const parseByDefaultBitcoinLikeBlockchainBalanceData = (response) => {
```
Name 'BitcoinLike' was used when we tried to make universal code for all bitcoin-like blockchains...

ozalexo commented on src/navigation/WalletStack.js:
```diff
@@ -55,7 +55,7 @@ const WalletStack = createStackNavigator(
       navigationOptions: ({ navigation }) => ({
         title: i18n.t('ScreensTitles.Wallet'),
-        headerLeft: <HeaderBackButton onPress={() => navigation.goBack()} tintColor='white' />,
+        headerLeft: <HeaderBackButton onPress={() => navigation.navigate('WalletList', { updated: true })} tintColor='white' />,
```
Could you please briefly describe an idea, why we need 'updated' here?

ozalexo commented on src/screens/Wallet/Wallet/WalletContainer.js:
```diff
@@ -5,18 +5,21 @@
   }
 }
 
+const mapDispatchToProps = (dispatch) => bindActionCreators({}, dispatch)
```
Not sure that we need empty mapDispatchToProps

ozalexo commented on src/screens/Wallet/Wallet/WalletInfo/WalletInfoStyles.js:
```diff
@@ -19,7 +19,7 @@ export default StyleSheet.create({
   },
   walletDetailsSection: {
-    backgroundColor: colors.textContainer,
+    // backgroundColor: colors.textContainer,
```
Is it develop change or this style may be removed?

e-molochko replied:
With this bg color this screen looks different from others. With it it looks filled with color, not "transparent" like other

ozalexo commented on src/screens/Wallet/WalletList/WalletList.js:
```diff
@@ -53,20 +56,27 @@ export default class WalletList extends PureComponent {
 
 WalletList.propTypes = {
-  navigate: PropTypes.func,
+  navigation: PropTypes.shape({}),
```
It seem incorrect.  Please copy-paste it from any other file.

molochkoego committed:
PR comments changes

molochkoego committed:
temp, wortking with tx creation

ozalexo committed:
WIP: code ok

molochkoego committed:
Added signed/unsigned tx(temp)

ozalexo commented on src/screens/Wallet/Send/SendContainer.js:
```diff
@@ -120,6 +134,44 @@ class SendContainer extends React.Component {
+      }
+
+      requestBitcoinUtxoByAddress(address)
+        .then((results) => {
```
Below we need `.catch()`, even if it will contain `console.log`

ozalexo commented on packages/bitcoin/utils/amount.js:
```diff
@@ -20,6 +21,14 @@ export const parseByDefaultBitcoinLikeBlockchainBalanceData = (response) => {
+  return new BigNumber(BTC * DECIMALS)
+}
+
 export const convertToWei = (amount) => web3utils.toWei(amount)
```
It seems like here we will need try..catch. In the future, not now.

ozalexo commented on src/screens/Wallet/Send/Send.js:
```diff
@@ -40,11 +126,80 @@ export default class Send extends PureComponent {
         <PrimaryButton
           label='Go to password entrance'
           onPress={onTogglePasswordModal}
+          style={styles.openModalButton}
```
This button must be in the header. See https://test.chronobank.io/iOS/send.html
Its title should be "Done"

ozalexo commented on src/screens/Wallet/Send/SendContainer.js:
```diff
@@ -8,26 +8,385 @@ import {
+        value.startsWith('0x')
+
+      // Check for BitCoin-based
+      if (blockchain !== DUCK_ETHEREUM) {
```
DUCK_ETHEREUM must not be used like this. Here we need blockchain's name BLOCKCHAIN_ETHEREUM from  packages/ethereum/constants.js

ozalexo reviewed (commented):
I'm on Send screen in app, but there is no txDraft in the redux store. When it should appear?
![screenshot 2018-12-13 at 13 07 03](https://user-images.githubusercontent.com/661889/49919135-1ac04780-fed8-11e8-93e4-7a4a2ed1262e.png)

ozalexo committed:
Fixed import accounts.  Fixed Promise of RMQ connection (now it waits when really connected)

e-molochko committed:
Merge pull request #1 from ozalexo/feature/work

Fixe RMQ, automatic network connect, fixed import account, reworked login

molochkoego committed:
Selected network fixes + PR comments fixes

molochkoego committed:
Added signing by bio from modal

molochkoego committed:
Added work with store for tx
- Added biometry sign of tx
- Added balance for btc wallets on wallets list

molochkoego committed:
PR conflicts fixes

ozalexo committed:
Fixed import formik. Rework RMQ manager (work without debug) (#2)

Need to test on Android

molochkoego committed:
Little fixes

molochkoego committed:
Balance show for btc and wallets sections titles fixes

molochkoego committed:
Added tx send

molochkoego committed:
Added confirmSend modal + little fixes

molochkoego committed:
Fixed navigation after send

molochkoego committed:
Fixed fee calculation for BTC

molochkoego committed:
Fixed transaction send

ozalexo committed:
Fixed BTC creation (parent address)

ozalexo committed:
Fixed sections. Added debugs for weallet create process.

ozalexo committed:
Fixed wallets generation (now derived privateKey is using)

ozalexo committed:
WIP: fixing modals (does not work)

molochkoego committed:
Modal fixes

molochkoego committed:
Modal styles fixes

ozalexo committed:
Moved "Done" button to header

ozalexo committed:
Minor fix
